services/google_scraping_service/parser.go

```diff
@@ -0,0 +1,97 @@
+package google_scraping_service
+
+import (
+	"net/http"
+
+	"github.com/PuerkitoBio/goquery"
+)
+
+type GoogleResponseParser struct {
+	GoogleResponse *http.Response
+}
+
+type ParsingResult struct {
+	HtmlCode               string
+	LinksCount             int
+	NonAdwordsCount        int
+	NonAdwordLinks         []string
+	TopPostionAdwordsCount int
+	TopPositionAdwordLinks []string
+	TotalAdwordsCount      int
+}
+
+var ParseGoogleResponse = func(googleResp *http.Response) (ParsingResult, error) {
+	countBottomPositionAdwordsCh := make(chan int)
+	countLinksCh := make(chan int)
+	countNonAdwordsCh := make(chan int)
+	countTopPositionAdwordsCh := make(chan int)
+
+	fetchNonAdwordLinksCh := make(chan []string)
+	fetchTopPositionAdwordLinksCh := make(chan []string)
+
+	doc, err := goquery.NewDocumentFromReader(googleResp.Body)
+	if err != nil {
+		return ParsingResult{}, err
+	}
+
+	go countBottomPositionAdwords(doc, countBottomPositionAdwordsCh)
+	go countLinks(doc, countLinksCh)
+	go countNonAdwords(doc, countNonAdwordsCh)
+	go countTopPositionAdwords(doc, countTopPositionAdwordsCh)
+
+	go fetchNonAdwordLinks(doc, fetchNonAdwordLinksCh)
+	go fetchTopPositionAdwordLinks(doc, fetchTopPositionAdwordLinksCh)
+
+	bottomPositionAdwordsCount := <-countBottomPositionAdwordsCh
```

Comment:
I think we could improve the performance at this place by using [select](https://tour.golang.org/concurrency/5) so we can have the concurrency process here, with the current implementation, the result comes in sequential, it needs to wait for `<- countBottomPositionAdwordsCh` has the value before going to `<- countTopPositionAdwordsCh`, same for other chans bellow

Reply:
I also would be interested to check if that changes anything here 😁

Reply:
Tried to use `select` in 297b97c8c20002c290d3c2ac810769c9b841f609